polyply/src/big_smile_parsing.py

```diff
@@ -0,0 +1,275 @@
+            prev_node = current
+            current += 1
+
+        # terminate branch and jump back to anchor
+        branch_stop = _find_next_character(pattern, ['['], stop) >\
+                      _find_next_character(pattern, [')'], stop)
+        if stop <= len(pattern) and branch_stop and branching:
+            branching = False
+            prev_node = branch_anchor
+            # we have to multiply the branch n-times
+            eon_a = _find_next_character(pattern, [')'], stop)
+            if stop+1 < len(pattern) and pattern[eon_a+1] == "|":
+                eon_b = _find_next_character(pattern, ['[', ')', '(', '}'], eon_a+1)
+                # -1 because one branch has already been added at this point
+                for _ in range(0,int(pattern[eon_a+2:eon_b])-1):
+                    for bdx, (resname, n_mon) in enumerate(recipie):
+                        if bdx == 0:
+                            anchor = current
+                        for _ in range(0, n_mon):
+                            connection = [(prev_node, current)]
+                            meta_mol.add_monomer(current,
+                                                 resname,
+                                                 connection)
+                            prev_node = current
+                            current += 1
+                    prev_node = anchor
+    return meta_mol
+
+def _big_smile_iter(smile):
+    for token in smile:
+        yield token
+
+def tokenize_big_smile(big_smile):
+    """
+    Processes a BigSmile string by storing the
+    the BigSmile specific bonding descriptors
+    in a dict with refernce to the atom they
+    refer to. Furthermore, a cleaned smile
+    string is generated with the BigSmile
+    specific syntax removed.
+
+    Parameters
+    ----------
+    smile: str
+        a BigSmile smile string
+
+    Returns
+    -------
+    str
+        a canonical smile string
+    dict
+        a dict mapping bonding descriptors
+        to the nodes within the smile
+    """
+    smile_iter = _big_smile_iter(big_smile)
+    bonding_descrpt = defaultdict(list)
+    smile = ""
+    node_count = 0
+    prev_node = 0
+    for token in smile_iter:
+        if token == '[':
+            peek = next(smile_iter)
+            if peek in ['$', '>', '<']:
+                bond_descrp = peek
+                peek = next(smile_iter)
+                while peek != ']':
+                    bond_descrp += peek
+                    peek = next(smile_iter)
+                bonding_descrpt[prev_node].append(bond_descrp)
+            else:
+                smile = smile + token + peek
+                prev_node = node_count
+                node_count += 1
+
+        elif token == '(':
+            anchor = prev_node
+            smile += token
+        elif token == ')':
+            prev_node = anchor
+            smile += token
+        else:
+            if token not in '@ . - = # $ : / \\ + - %'\
```

Comment:
nit. Otherwise space is also a valid character

```suggestion
            if token not in '@ . - = # $ : / \\ + - %'.split()\
```

Reply:
I don't see the advantage?